### ISSUE TITLE
BFGS always failed but Adam works

### ISSUE THREAD

**Issue by fmc202** [pyro-ppl/numpyro#999]
I was trying to fit a very simple linear model from Statistical rethinking using SVI with autolaplaceapproximation. However, the adam optimizer works fine but the BFGS gives NAN values in loss. Also, the model fitted using BFGS have singular Hessian at MAP point. Please see code below:

```python
import jax.numpy as jnp
from sklearn.preprocessing import StandardScaler
import jax
import numpyro
import matplotlib.pyplot as plt
from numpyro.infer import SVI, Trace_ELBO,Predictive
from numpyro.infer.autoguide import AutoLaplaceApproximation,AutoDelta
import arviz as az
import pandas as pd
from numpyro.diagnostics import hpdi, print_summary

dist = numpyro.distributions
numpyro.enable_x64()

data = pd.read_csv("./data/Howell1.csv",sep=";")
weight = data["weight"].to_numpy()
height = data["height"].to_numpy()
scaler = StandardScaler()
scaler.fit(weight.reshape([-1,1]))
weight_transform = scaler.transform(weight.reshape([-1,1])).flatten()

def model(x=None,x2=None,h=None):
    a = numpyro.sample("a",dist.Normal(178,20))
    b1 = numpyro.sample("b1", dist.LogNormal(0, 1))
    b2 = numpyro.sample("b2", dist.Normal(0, 1))
    sigma = numpyro.sample("sigma",dist.Uniform(0,50))
    mu = numpyro.deterministic("mu",a + b1*x +b2*x2)
    numpyro.sample("obs",dist.Normal(mu,sigma),obs=h)

guide = AutoLaplaceApproximation(model)
optimizer=numpyro.optim.Minimize()
svi = SVI(model, guide, optimizer, Trace_ELBO(),x=weight_transform,x2=weight_transform**2,h=height)
params, losses = svi.run(jax.random.PRNGKey(3), 3000)
samples = guide.sample_posterior(jax.random.PRNGKey(3), params, (1000,))
```
The results from BFGS:
100%|██████████| 3000/3000 [00:03<00:00, 767.80it/s, init loss: 2901.7748, avg. loss [2851-3000]: nan]

python3.8/site-packages/numpyro/infer/autoguide.py:764: UserWarning: Hessian of log posterior at the MAP point is singular. Posterior samples from AutoLaplaceApproxmiation will be constant (equal to the MAP point).
  warnings.warn("Hessian of log posterior at the MAP point is singular. Posterior"

        mean       std    median      5.5%     94.5%     n_eff     r_hat
         a       nan       nan       nan       nan       nan       nan       nan
        b1       nan       nan       nan       nan       nan       nan       nan
        b2       nan       nan       nan       nan       nan       nan       nan
     sigma       nan       nan       nan       nan       nan       nan       nan

results from adam:
         mean       std    median      5.5%     94.5%     n_eff     r_hat
         a    146.03      0.36    146.03    145.45    146.57    994.72      1.00
        b1     21.84      0.28     21.84     21.40     22.30   1014.59      1.00
        b2     -7.78      0.27     -7.78     -8.23     -7.38    988.45      1.00
     sigma      5.77      0.18      5.77      5.49      6.06   1036.88      1.00

**Comment by fehiepsi** [comment]
Hi @fmc202, for Minimize, you just need to set `num_steps` to `1`: `svi.run(jax.random.PRNGKey(3), 1)`.

By the way, Hessian matrix is not always positive definite. I guess we can open a feature request to compute Gauss-Newton Hessian matrix.

**Comment by fehiepsi** [comment]
I opened #1001 to seek help with having a positive definite version of the Hessian matrix. Hope it will resolve the issue. :)

**Comment by ColdTeapot273K** [comment]
In case anyone encounters this problem and their AutoLaplaceApproximation + Adam gives this error, trying lower stepsize for Adam or different optimizers can help aswell (because substituting guide to AutoNormal can give weird results on some problems)
